src/expr/core/src/window_function/call.rs

```diff
@@ -154,6 +154,40 @@ impl RowsFrameBounds {
     fn validate(&self) -> Result<()> {
         FrameBound::validate_bounds(&self.start, &self.end)
     }
+
+    /// Check if the `ROWS` frame is canonical.
+    ///
+    /// A canonical `ROWS` frame is defined as:
+    ///
+    /// - It is valid.
+    /// - It is a crossing frame, i.e. it contains the current row.
+    pub fn is_canonical(&self) -> bool {
+        self.validate().is_ok() && {
+            let start = self.start.to_offset();
+            let end = self.end.to_offset();
+            start.unwrap_or(0) <= 0 && end.unwrap_or(0) >= 0
+        }
+    }
+
+    /// Get the number of preceding rows.
+    pub fn n_preceding_rows(&self) -> Option<usize> {
+        match (&self.start, &self.end) {
+            (UnboundedPreceding, _) => None,
+            (Preceding(n1), Preceding(n2)) => Some(*n1.max(n2)),
```

Comment:
IIUC, n1 should always be greater than n2.

e.g. n1: 2, n2: 1.

perhaps we should add this assertion here:
```
n1 >= n2
```

Then select n2 to be n preceding rows.

Reply:
`n1` can be greater than `n2`, `ROWS BETWEEN 8 PRECEDING AND 2 PRECEDING` is a valid frame definition😐

Reply:
Actually the algo here is conservative. I'm afraid we miss some very corner case so just use the *most preceding* bound and *most following* bound, although the frame may contain nothing which means the *most preceding* (e.g. `8 PRECEDING` in the above comment) may be never touched. In the conservative algorithm we will include it in the affected range. This won't hurt the correctness when it comes to function result calculation.